File: llmio-node/src/services/rpm-limiter.ts
/**
 * RPM（每分钟请求数）限流服务
 * 使用滑动窗口算法实现精确的分钟级限流
 */

import type { Redis } from "ioredis";

// 内存存储（当 Redis 不可用时使用）
const memoryStore: Map<string, number[]> = new Map();

// Redis 实例（可选）
let redisClient: Redis | null = null;

/**
 * 初始化 RPM 限流器
 * @param redis Redis 客户端（可选）
 */
export function initRpmLimiter(redis: Redis | null) {
  redisClient = redis;
}

/**
 * 获取供应商的 RPM 计数键
 */
function getRpmKey(providerId: number): string {
  return `rpm:provider:${providerId}`;
}

/**
 * 检查供应商是否达到 RPM 限制
 * @param providerId 供应商 ID
 * @param rpmLimit RPM 限制（0 表示无限制）
 * @returns 是否允许请求
 */
export async function checkRpmLimit(
  providerId: number,
  rpmLimit: number
): Promise<boolean> {
  // 0 表示无限制
  if (rpmLimit <= 0) {
    return true;
  }

  const now = Date.now();
  const windowStart = now - 60 * 1000; // 1分钟前

  if (redisClient) {
    return await checkRpmLimitRedis(providerId, rpmLimit, now, windowStart);
  } else {
    return checkRpmLimitMemory(providerId, rpmLimit, now, windowStart);
  }
}

/**
 * 记录一次请求（增加计数）
 * @param providerId 供应商 ID
 */
export async function recordRpmRequest(providerId: number): Promise<void> {
  const now = Date.now();

  if (redisClient) {
    await recordRpmRequestRedis(providerId, now);
  } else {
    recordRpmRequestMemory(providerId, now);
  }
}

/**
 * 获取供应商当前的 RPM 计数
 * @param providerId 供应商 ID
 * @returns 当前分钟内的请求数
 */
export async function getCurrentRpmCount(providerId: number): Promise<number> {
  const now = Date.now();
  const windowStart = now - 60 * 1000;

  if (redisClient) {
    return await getCurrentRpmCountRedis(providerId, windowStart, now);
  } else {
    return getCurrentRpmCountMemory(providerId, windowStart);
  }
}

// ==================== Redis 实现 ====================

async function checkRpmLimitRedis(
  providerId: number,
  rpmLimit: number,
  now: number,
  windowStart: number
): Promise<boolean> {
  const key = getRpmKey(providerId);

  try {
    // 移除过期的请求记录
    await redisClient!.zremrangebyscore(key, 0, windowStart);

    // 获取当前窗口内的请求数
    const count = await redisClient!.zcard(key);

    return count < rpmLimit;
  } catch (e) {
    console.error(`[rpm-limiter] Redis check failed for provider ${providerId}:`, e);
    // Redis 出错时降级为允许
    return true;
  }
}

async function recordRpmRequestRedis(
  providerId: number,
  now: number
): Promise<void> {
  const key = getRpmKey(providerId);

  try {
    // 使用有序集合存储请求时间戳
    // score 和 member 都使用时间戳，确保唯一性
    await redisClient!.zadd(key, now, `${now}-${Math.random()}`);

    // 设置过期时间为 2 分钟（确保数据能被清理）
    await redisClient!.expire(key, 120);
  } catch (e) {
    console.error(`[rpm-limiter] Redis record failed for provider ${providerId}:`, e);
  }
}

async function getCurrentRpmCountRedis(
  providerId: number,
  windowStart: number,
  now: number
): Promise<number> {
  const key = getRpmKey(providerId);

  try {
    // 移除过期的请求记录
    await redisClient!.zremrangebyscore(key, 0, windowStart);

    // 获取当前窗口内的请求数
    return await redisClient!.zcard(key);
  } catch (e) {
    console.error(`[rpm-limiter] Redis count failed for provider ${providerId}:`, e);
    return 0;
  }
}

// ==================== 内存实现 ====================

function checkRpmLimitMemory(
  providerId: number,
  rpmLimit: number,
  _now: number,
  windowStart: number
): boolean {
  const key = getRpmKey(providerId);
  const timestamps = memoryStore.get(key) || [];

  // 过滤出窗口内的请求
  const validTimestamps = timestamps.filter((t) => t > windowStart);

  // 更新存储
  memoryStore.set(key, validTimestamps);

  return validTimestamps.length < rpmLimit;
}

function recordRpmRequestMemory(providerId: number, now: number): void {
  const key = getRpmKey(providerId);
  const timestamps = memoryStore.get(key) || [];

  // 添加新的时间戳
  timestamps.push(now);

  // 清理过期数据（保留最近2分钟的）
  const windowStart = now - 120 * 1000;
  const validTimestamps = timestamps.filter((t) => t > windowStart);

  memoryStore.set(key, validTimestamps);
}

function getCurrentRpmCountMemory(
  providerId: number,
  windowStart: number
): number {
  const key = getRpmKey(providerId);
  const timestamps = memoryStore.get(key) || [];

  // 过滤出窗口内的请求
  const validTimestamps = timestamps.filter((t) => t > windowStart);

  return validTimestamps.length;
}

/**
 * 清理所有内存中的 RPM 数据（用于测试）
 */
export function clearRpmData(): void {
  memoryStore.clear();
}

File: limiter/rpm_limiter.go
package limiter

import (
	"context"
	"fmt"
	"strconv"
	"sync"
	"time"

	"github.com/go-redis/redis/v8"
)

// RPMLimiter RPM限流器
type RPMLimiter struct {
	redis  *redis.Client
	memory *sync.Map // 内存存储，当Redis不可用时使用
}

// RequestRecord 请求记录
type RequestRecord struct {
	Timestamps []int64 `json:"timestamps"`
}

// NewRPMLimiter 创建新的RPM限流器
func NewRPMLimiter(redisClient *redis.Client) *RPMLimiter {
	return &RPMLimiter{
		redis:  redisClient,
		memory: &sync.Map{},
	}
}

// CheckRPMLimit 检查是否达到RPM限制
func (r *RPMLimiter) CheckRPMLimit(ctx context.Context, providerID uint, rpmLimit int) (bool, error) {
	// 0表示无限制
	if rpmLimit <= 0 {
		return true, nil
	}

	now := time.Now().Unix()
	windowStart := now - 60 // 1分钟前

	if r.redis != nil {
		return r.checkRPMLimitRedis(ctx, providerID, rpmLimit, now, windowStart)
	}

	return r.checkRPMLimitMemory(providerID, rpmLimit, now, windowStart), nil
}

// RecordRequest 记录一次请求
func (r *RPMLimiter) RecordRequest(ctx context.Context, providerID uint) error {
	now := time.Now().Unix()

	if r.redis != nil {
		return r.recordRequestRedis(ctx, providerID, now)
	}

	r.recordRequestMemory(providerID, now)
	return nil
}

// GetCurrentRPMCount 获取当前RPM计数
func (r *RPMLimiter) GetCurrentRPMCount(ctx context.Context, providerID uint) (int, error) {
	now := time.Now().Unix()
	windowStart := now - 60

	if r.redis != nil {
		return r.getCurrentRPMCountRedis(ctx, providerID, windowStart, now)
	}

	return r.getCurrentRPMCountMemory(providerID, windowStart), nil
}

// getRPMKey 获取RPM存储键
func (r *RPMLimiter) getRPMKey(providerID uint) string {
	return fmt.Sprintf("rpm:provider:%d", providerID)
}

// ==================== Redis实现 ====================

func (r *RPMLimiter) checkRPMLimitRedis(ctx context.Context, providerID uint, rpmLimit int, now, windowStart int64) (bool, error) {
	key := r.getRPMKey(providerID)

	// 使用Redis事务确保原子性
	pipe := r.redis.TxPipeline()

	// 移除过期的请求记录
	pipe.ZRemRangeByScore(ctx, key, "0", strconv.FormatInt(windowStart, 10))

	// 获取当前窗口内的请求数
	countCmd := pipe.ZCard(ctx, key)

	_, err := pipe.Exec(ctx)
	if err != nil {
		// Redis出错时降级为允许
		return true, fmt.Errorf("redis rpm check failed: %w", err)
	}

	count := countCmd.Val()
	return count < int64(rpmLimit), nil
}

func (r *RPMLimiter) recordRequestRedis(ctx context.Context, providerID uint, now int64) error {
	key := r.getRPMKey(providerID)

	// 使用有序集合存储请求时间戳
	// score和member都使用时间戳，确保唯一性
	member := fmt.Sprintf("%d-%d", now, time.Now().UnixNano()%1000000)

	pipe := r.redis.TxPipeline()
	pipe.ZAdd(ctx, key, &redis.Z{
		Score:  float64(now),
		Member: member,
	})

	// 设置过期时间为2分钟
	pipe.Expire(ctx, key, 2*time.Minute)

	_, err := pipe.Exec(ctx)
	return err
}

func (r *RPMLimiter) getCurrentRPMCountRedis(ctx context.Context, providerID uint, windowStart, now int64) (int, error) {
	key := r.getRPMKey(providerID)

	pipe := r.redis.TxPipeline()

	// 移除过期的请求记录
	pipe.ZRemRangeByScore(ctx, key, "0", strconv.FormatInt(windowStart, 10))

	// 获取当前窗口内的请求数
	countCmd := pipe.ZCard(ctx, key)

	_, err := pipe.Exec(ctx)
	if err != nil {
		return 0, err
	}

	return int(countCmd.Val()), nil
}

// ==================== 内存实现 ====================

func (r *RPMLimiter) checkRPMLimitMemory(providerID uint, rpmLimit int, now, windowStart int64) bool {
	key := r.getRPMKey(providerID)

	value, exists := r.memory.Load(key)
	if !exists {
		return true
	}

	record, ok := value.(*RequestRecord)
	if !ok {
		return true
	}

	// 过滤出窗口内的请求
	validTimestamps := make([]int64, 0, len(record.Timestamps))
	for _, ts := range record.Timestamps {
		if ts > windowStart {
			validTimestamps = append(validTimestamps, ts)
		}
	}

	// 更新存储
	record.Timestamps = validTimestamps
	r.memory.Store(key, record)

	return len(validTimestamps) < rpmLimit
}

func (r *RPMLimiter) recordRequestMemory(providerID uint, now int64) {
	key := r.getRPMKey(providerID)

	value, exists := r.memory.Load(key)
	var record *RequestRecord

	if exists {
		record, _ = value.(*RequestRecord)
	}

	if record == nil {
		record = &RequestRecord{
			Timestamps: make([]int64, 0),
		}
	}

	// 添加新的时间戳
	record.Timestamps = append(record.Timestamps, now)

	// 清理过期数据（保留最近2分钟的）
	windowStart := now - 120
	validTimestamps := make([]int64, 0, len(record.Timestamps))
	for _, ts := range record.Timestamps {
		if ts > windowStart {
			validTimestamps = append(validTimestamps, ts)
		}
	}

	record.Timestamps = validTimestamps
	r.memory.Store(key, record)
}

func (r *RPMLimiter) getCurrentRPMCountMemory(providerID uint, windowStart int64) int {
	key := r.getRPMKey(providerID)

	value, exists := r.memory.Load(key)
	if !exists {
		return 0
	}

	record, ok := value.(*RequestRecord)
	if !ok {
		return 0
	}

	// 过滤出窗口内的请求
	count := 0
	for _, ts := range record.Timestamps {
		if ts > windowStart {
			count++
		}
	}

	return count
}

// ClearMemoryData 清理内存数据（用于测试）
func (r *RPMLimiter) ClearMemoryData() {
	r.memory = &sync.Map{}
}

// GetStats 获取限流器统计信息
func (r *RPMLimiter) GetStats(ctx context.Context) map[string]interface{} {
	stats := map[string]interface{}{
		"storage_type": "memory",
		"providers":    make(map[string]int),
	}

	if r.redis != nil {
		stats["storage_type"] = "redis"

		// 获取Redis中的RPM键
		keys, err := r.redis.Keys(ctx, "rpm:provider:*").Result()
		if err == nil {
			providerStats := make(map[string]int)
			for _, key := range keys {
				count, err := r.redis.ZCard(ctx, key).Result()
				if err == nil {
					providerStats[key] = int(count)
				}
			}
			stats["providers"] = providerStats
		}
	} else {
		// 内存统计
		providerStats := make(map[string]int)
		r.memory.Range(func(key, value interface{}) bool {
			if keyStr, ok := key.(string); ok {
				if record, ok := value.(*RequestRecord); ok {
					providerStats[keyStr] = len(record.Timestamps)
				}
			}
			return true
		})
		stats["providers"] = providerStats
	}

	return stats
}
